### PR TITLE
Support Git for Windows 2.x

### PR DESCRIPTION
**Pull request by dscho**
This is the big one. This is the Pull Request modifying mostly the path conversion handling so that Git for Windows 2.x is backwards compatible to Git for Windows 1.x, and more importantly: so that it passes the test suite of Git proper.

Please note that I am prepared to make changes, so if there is anything you do not like, please state explicitly what you would like _instead_. Thank you.

Let's get this done.

### CODE REVIEW

**Comment by mingwandroid** [comment]
Hi @dscho we'll get it done sure, but I hope everyone is suitably patient. I'm off on holiday in a few weeks and have quite a backlog of things to do, but don't be disheartened by this, it's a very important merge request.

**Comment by dscho** [comment]
@mingwandroid yeah, I expected this to take some time, and also effort from my side. But I think that we're in a good position now: Git for Windows ships with its own slightly modified msys2-runtime/mintty, so there is no rush, we can go about integrating the changes leisurely, and more importantly: in a way that MSys2 *and* Git for Windows are happy about.

**Comment by dscho** [comment]
I just rebased onto `msys2-master`.

**Comment by Alexpux** [comment]
Modifying path conversion procedure breal many tests:
````
warlock+alexx@WARLOCK MSYS ~/path_convert
$ PATH=/mingw32/bin:$PATH ./build.sh
test 0 passed: src="-LIBPATH:../lib", dst="-LIBPATH:../lib"
test 1 failed: src="-LIBPATH:../lib:/tmp", dst="-LIBPATH:../lib:/tmp" expect="-LIBPATH:..\lib;C:\building\msys32\tmp"
test 2 passed: src="//Collection:http://tfsserver", dst="//Collection:http://tfsserver"
test 3 passed: src="/Collection:http://tfsserver", dst="/Collection:http://tfsserver"
test 4 failed: src="-L'/foo bar/boo' PREFIX='/foo bar/boo'", dst="-L'/foo bar/boo' PREFIX='/foo bar/boo'" expect="-L'C:/building/msys32/foo bar/boo' PREFIX='/foo bar/boo'"
test 5 failed: src="-L'/foo /bar/boo' PREFIX='/foo /bar/boo'", dst="-L'/foo /bar/boo' PREFIX='/foo /bar/boo'" expect="-L'C:/building/msys32/foo /bar/boo' PREFIX='/foo /bar/boo'"
test 6 passed: src="C:\foo\bar", dst="C:\foo\bar"
test 7 passed: src="/foo/bar;", dst="/foo/bar;"
test 8 passed: src="//foobar", dst="/foobar"
test 9 passed: src="//foo\bar", dst="/foo/bar"
test 10 passed: src="//foo/bar", dst="//foo/bar"
test 11 passed: src="/c:\foo\bar", dst="c:/foo/bar"
test 12 failed: src="foo=/bar", dst="foo=C:/msys32/bar" expect="foo=C:/building/msys32/bar"
test 13 failed: src="-foo,/bar", dst="-foo,C:/msys32/bar" expect="-foo,C:/building/msys32/bar"
test 14 failed: src="-I/foo,/bar", dst="-I/foo,C:/msys32/bar" expect="-I/foo,C:/building/msys32/bar"
test 15 failed: src="-I/foo", dst="-IC:/msys32/foo" expect="-IC:/building/msys32/foo"
test 16 failed: src="-L/foo", dst="-LC:/msys32/foo" expect="-LC:/building/msys32/foo"
test 17 failed: src="-L'/foo /bar'", dst="-L'/foo /bar'" expect="-L'C:/building/msys32/foo /bar'"
test 18 failed: src="-L'/foo bar'", dst="-L'/foo bar'" expect="-L'C:/building/msys32/foo bar'"
test 19 failed: src="'/opt /bin'", dst="'/opt /bin'" expect="'C:/building/msys32/opt /bin'"
test 20 failed: src="'/opt files/bin'", dst="'/opt files/bin'" expect="'C:/building/msys32/opt files/bin'"
test 21 failed: src="/", dst="C:/msys32/" expect="C:/building/msys32/"
test 22 passed: src="/..", dst="/.."
test 23 passed: src="x:x:/x", dst="x:x:/x"
test 24 failed: src="x::x:/xx", dst="x::x:/xx" expect="x;x;C:\building\msys32\xx"
test 25 failed: src="x::x/z:x", dst="x::x/z:x" expect="x;x\z;x"
test 26 failed: src="x::/x z:x", dst="x::/x z:x" expect="x;C:\msys32\x z;x"
test 27 failed: src="'x::/x z:x'", dst="'x::/x z:x'" expect="'x;C:\msys32\x z;x'"
test 28 passed: src="/dev/null", dst="nul"
test 29 failed: src="'/dev/null'", dst="'/dev/null'" expect="'nul'"
test 30 failed: src="/tmp:/tmp", dst="C:\msys32\tmp;C:\msys32\tmp" expect="C:\building\msys32\tmp;C:\building\msys32\tmp"
test 31 failed: src="'/tmp:/tmp'", dst="'/tmp:/tmp'" expect="'C:\building\msys32\tmp;C:\building\msys32\tmp'"
test 32 failed: src="-L'/tmp:/tmp'", dst="-L'/tmp:/tmp'" expect="-L'C:\building\msys32\tmp;C:\building\msys32\tmp'"
test 33 failed: src="-L/tmp:/tmp", dst="-LC:\msys32\tmp;C:\msys32\tmp" expect="-LC:\building\msys32\tmp;C:\building\msys32\tmp"
test 34 failed: src="'/bin:/Program Files:/lib'", dst="'/bin:/Program Files:/lib'" expect="'C:\building\msys32\usr\bin;C:\building\msys32\Program Files;C:\building\msys32\lib'"
test 35 failed: src="'-L/opt /bin'", dst="'-L/opt /bin'" expect="'-LC:/building/msys32/opt /bin'"
test 36 passed: src="-w -- INSTALL_ROOT=C:/Test/ports64", dst="-w -- INSTALL_ROOT=C:/Test/ports64"
test 37 passed: src="-w -- INSTALL_ROOT=C:\Test\ports64", dst="-w -- INSTALL_ROOT=C:\Test\ports64"
test 38 passed: src="-IC:/Test/ports64", dst="-IC:/Test/ports64"
test 39 passed: src="'C:\foo\bar'", dst="'C:\foo\bar'"
test 40 passed: src="'/foo/bar;'", dst="'/foo/bar;'"
test 41 failed: src="'//foobar'", dst="'//foobar'" expect="'/foobar'"
test 42 failed: src="'//foo\bar'", dst="'//foo\bar'" expect="'/foo/bar'"
test 43 passed: src="'//foo/bar'", dst="'//foo/bar'"
test 44 failed: src="'/c:\foo\bar'", dst="'/c:\foo\bar'" expect="'c:/foo/bar'"
test 45 failed: src="'foo=/bar'", dst="'foo=/bar'" expect="'foo=C:/building/msys32/bar'"
test 46 failed: src="'-foo,/bar'", dst="'-foo,/bar'" expect="'-foo,C:/building/msys32/bar'"
test 47 failed: src="'-I/foo,/bar'", dst="'-I/foo,/bar'" expect="'-I/foo,C:/building/msys32/bar'"
test 48 f./build.sh: line 7:  4760 Segmentation fault      ./path_conv_test.exe

NOT GOOD
````
So in current state I can't aceept this patch because it break primary use of MSYS2 - building MINGW software.

@dscho , here is our repository for path conversion procedure testing:
https://github.com/Alexpux/path_convert

File: winsup/cygwin/path.cc [hunk 1]
```diff
@@ -734,11 +734,11 @@ path_conv::check (const char *src, unsigned opt,
 	      need_directory = 1;
 	      *--tail = '\0';
 	    }
-	  /* Special case for "/" must set need_directory, without removing
+	  /* Special case for "/" must not set need_directory, neither remove
 	     trailing slash */
 	  else if (tail == path_copy + 1 && isslash (tail[-1]))
 	    {
-	      need_directory = 1;
+	      need_directory = 0;
```
Comment (Alexpux) [inline winsup/cygwin/path.cc]: @mingwandroid, this is what you fix and @dscho want to revert.

Comment (mingwandroid) [inline winsup/cygwin/path.cc]: Why revert this? It's pretty important. For example you can't build a sysrooted GCC without it.

Comment (mingwandroid) [inline winsup/cygwin/path.cc]: And logically, of the input directory ended / then the output should also.

Comment (dscho) [inline winsup/cygwin/path.cc]: Heh, this is a fix. As in: we're in a fix now... because the trailing slash breaks Git for Windows pretty badly. See https://github.com/dscho/msys2-runtime/commit/560d5675a735f2f917f49b7411a9fa01b87fa200 (The test in question is one of Git' extensive tests: https://github.com/git/git/blob/master/t/t0060-path-utils.sh)

Comment (mingwandroid) [inline winsup/cygwin/path.cc]: If I read this correctly, the test case I'd relying on testing old behaviour in MSYS? If so I'd argue that the test is broken. I've not followed the chain of where normalize_path ends up at before getting into cygwin land, can you key Mr know where that is?

My reasoning for this change are pretty sound - even if cygwin dismissed it out if hand without a reasonable explanation.

AFAICR, before my change, the conv_path function would retain any final / for folders that existed but wouldn't pay the same courtesy for those that didn't. I can see no reason for such unexpected behaviour!

Comment (mingwandroid) [inline winsup/cygwin/path.cc]: Urgh, autocorrect hell

Comment (dscho) [inline winsup/cygwin/path.cc]: I will run the test with and without this patch again as soon as I have taken care of some day-job issues and will get back to you.

Comment (dscho) [inline winsup/cygwin/path.cc]: > And logically, if the input directory ended / then the output should also.

This is only half-true. For every directory, you can specify the path with, and without a trailing slash. For every directory except for `/`, i.e. the POSIX root directory, that is.

So here is the crux: Git expects you to be able to pass directories *without* a trailing slash (except for "/"). It does test this several times, and if any directory other than / arrives (and "/" cannot arrive because we are mangling that path), the unit tests fail. In total, 16 test cases **fail** in t0060 if you cannot pass `/` as a directory *without* a trailing slash after POSIX-to-Windows path mangling. And these are just the parts in t0060 that have this -- very valid -- expectation.

You can argue both ways, of course: since `/` already "ended in a slash", it should produce a `C:\msys64\` *with* a trailing slash. However, the MinGW program that received that path never knew that it was `/` and can legally expect it *not* to end in a trailing slash.

"Fixing" the path normalization in Git to *always* strip trailing slashes is inappropriate, of course. And t0060 reflects that, by failing 10 *other* test cases (although all 16 original ones would be "fixed").

We will need some solution there to allow us to configure *how* MSys2 treats that corner case rather than forcing one world view on everybody, I think.

Comment (dscho) [inline winsup/cygwin/path.cc]: @mingwandroid would you be okay with a setting, say, in `nsswitch.conf` to configure whether to add or not to add that slash?

Comment (dscho) [inline winsup/cygwin/path.cc]: @mingwandroid or maybe an environment variable to toggle the behavior, defaulting to the behavior you prefer?

Comment (mingwandroid) [inline winsup/cygwin/path.cc]: Hi Johannes. I'm on holiday and without internet access for the next 2
weeks (I'm typing this from an internet cafe). I'll get back to you on this
issue when I'm back. Sorry about the delay.

Comment (dscho) [inline winsup/cygwin/path.cc]: I just rebased the branch locally and tried to revert just this one patch. No joy:

```
not ok 51 - longest ancestor: /foo / => 14
#       actual=$(test-path-utils longest_ancestor_length '/foo' '/') &&
#                test "$actual" = '14'
ok 52 - longest ancestor: /foo /fo => -1
ok 53 - longest ancestor: /foo /foo => -1
ok 54 - longest ancestor: /foo /bar => -1
ok 55 - longest ancestor: /foo /foo/bar => -1
ok 56 - longest ancestor: /foo /foo:/bar => -1
not ok 57 - longest ancestor: /foo /:/foo:/bar => 14
#       actual=$(test-path-utils longest_ancestor_length '/foo' '/:/foo:/bar') &&
#                test "$actual" = '14'
not ok 58 - longest ancestor: /foo /foo:/:/bar => 14
#       actual=$(test-path-utils longest_ancestor_length '/foo' '/foo:/:/bar') &&
#                test "$actual" = '14'
not ok 59 - longest ancestor: /foo /:/bar:/foo => 14
#       actual=$(test-path-utils longest_ancestor_length '/foo' '/:/bar:/foo') &&
#                test "$actual" = '14'
not ok 60 - longest ancestor: /foo/bar / => 14
#       actual=$(test-path-utils longest_ancestor_length '/foo/bar' '/') &&
#                test "$actual" = '14'
ok 61 - longest ancestor: /foo/bar /fo => -1
not ok 62 - longest ancestor: /foo/bar /foo => 18
#       actual=$(test-path-utils longest_ancestor_length '/foo/bar' '/foo') &&
#                test "$actual" = '18'
ok 63 - longest ancestor: /foo/bar /foo/ba => -1
not ok 64 - longest ancestor: /foo/bar /:/fo => 14
#       actual=$(test-path-utils longest_ancestor_length '/foo/bar' '/:/fo') &&
#                test "$actual" = '14'
not ok 65 - longest ancestor: /foo/bar /foo:/foo/ba => 18
#       actual=$(test-path-utils longest_ancestor_length '/foo/bar' '/foo:/foo/ba') &&
#                test "$actual" = '18'
ok 66 - longest ancestor: /foo/bar /bar => -1
ok 67 - longest ancestor: /foo/bar /fo => -1
not ok 68 - longest ancestor: /foo/bar /foo:/bar => 18
#       actual=$(test-path-utils longest_ancestor_length '/foo/bar' '/foo:/bar') &&
#                test "$actual" = '18'
not ok 69 - longest ancestor: /foo/bar /:/foo:/bar => 18
#       actual=$(test-path-utils longest_ancestor_length '/foo/bar' '/:/foo:/bar') &&
#                test "$actual" = '18'
not ok 70 - longest ancestor: /foo/bar /foo:/:/bar => 18
#       actual=$(test-path-utils longest_ancestor_length '/foo/bar' '/foo:/:/bar') &&
#                test "$actual" = '18'
not ok 71 - longest ancestor: /foo/bar /:/bar:/fo => 14
#       actual=$(test-path-utils longest_ancestor_length '/foo/bar' '/:/bar:/fo') &&
#                test "$actual" = '14'
not ok 72 - longest ancestor: /foo/bar /:/bar => 14
#       actual=$(test-path-utils longest_ancestor_length '/foo/bar' '/:/bar') &&
#                test "$actual" = '14'
not ok 73 - longest ancestor: /foo/bar /foo => 18
#       actual=$(test-path-utils longest_ancestor_length '/foo/bar' '/foo') &&
#                test "$actual" = '18'
not ok 74 - longest ancestor: /foo/bar /foo:/bar => 18
#       actual=$(test-path-utils longest_ancestor_length '/foo/bar' '/foo:/bar') &&
#                test "$actual" = '18'
[...]
# failed 15 among 152 test(s)
```

Comment (dscho) [inline winsup/cygwin/path.cc]: > I just rebased the branch locally

For the record, it is available here: https://github.com/dscho/msys2-runtime/tree/git-for-windows-rebased.

Comment (dscho) [inline winsup/cygwin/path.cc]: BTW the output I pasted is from Git's own [`t0060-path-utils.sh`](http://github.com/git-for-windows/git/blob/master/t/t0060-path-utils.sh).

**Comment by mingwandroid** [comment]
@dscho Ok, I'm going to take a look again. It's good to see a big push happening.

**Comment by mingwandroid** [comment]
@dscho Just a quick ping in-case you know the solution already (I am going to go out soon so won't continue investigating tonight), trying to build the x86_64 msys2-runtime in Git for Windows SDK pointing the `PKGBUILD` at `source=('msys2-runtime'::'git+https://github.com/dscho/msys2-runtime.git#branch=git-for-windows-rebased'` I ran into:

```
gcc -L/usr/src/MSYS2-packages/msys2-runtime/src/build-x86_64-pc-msys/x86_64-pc-msys/winsup/cygwin -isystem /usr/src/MSYS2-packages/msys2-runtime/src/msys2-runtime/winsup/cygwin/include -B/usr/src/MSYS2-packages/msys2-runtime/src/build-x86_64-pc-msys/x86_64-pc-msys/newlib/ -isystem /usr/src/MSYS2-packages/msys2-runtime/src/build-x86_64-pc-msys/x86_64-pc-msys/newlib/targ-include -isystem /usr/src/MSYS2-packages/msys2-runtime/src/msys2-runtime/newlib/libc/include    -I/usr/src/MSYS2-packages/msys2-runtime/src/msys2-runtime/winsup/cygwin/include -DPACKAGE_NAME=\"newlib\" -DPACKAGE_TARNAME=\"newlib\" -DPACKAGE_VERSION=\"2.3.0\" -DPACKAGE_STRING=\"newlib\ 2.3.0\" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -I. -I/usr/src/MSYS2-packages/msys2-runtime/src/msys2-runtime/newlib/libc/syscalls -DHAVE_OPENDIR -DHAVE_RENAME -DSIGNAL_PROVIDED -D_COMPILING_NEWLIB -DHAVE_BLKSIZE -DHAVE_FCNTL -DMALLOC_PROVIDED -fno-builtin      -g -O2 -pipe -ggdb -c -o lib_a-sysclose.o `test -f 'sysclose.c' || echo '/usr/src/MSYS2-packages/msys2-runtime/src/msys2-runtime/newlib/libc/syscalls/'`sysclose.c
In file included from E:/git-sdk-64/usr/src/MSYS2-packages/msys2-runtime/src/msys2-runtime/newlib/libc/include/sys/unistd.h:12:0,
                 from E:/git-sdk-64/usr/src/MSYS2-packages/msys2-runtime/src/msys2-runtime/newlib/libc/include/unistd.h:4,
                 from E:/git-sdk-64/usr/src/MSYS2-packages/msys2-runtime/src/msys2-runtime/newlib/libc/syscalls/sysclose.c:4:
E:/git-sdk-64/usr/src/MSYS2-packages/msys2-runtime/src/msys2-runtime/newlib/libc/include/sys/types.h:124:18: error: conflicting types for 'time_t'
 typedef _TIME_T_ time_t;
                  ^
In file included from E:/git-sdk-64/mingw64/x86_64-w64-mingw32/include/stddef.h:7:0,
                 from E:/git-sdk-64/mingw64/lib/gcc/x86_64-w64-mingw32/5.3.0/include/stddef.h:1,
                 from E:/git-sdk-64/usr/src/MSYS2-packages/msys2-runtime/src/msys2-runtime/newlib/libc/include/sys/reent.h:14,
                 from E:/git-sdk-64/usr/src/MSYS2-packages/msys2-runtime/src/msys2-runtime/newlib/libc/include/reent.h:93,
                 from E:/git-sdk-64/usr/src/MSYS2-packages/msys2-runtime/src/msys2-runtime/newlib/libc/syscalls/sysclose.c:3:
E:/git-sdk-64/mingw64/x86_64-w64-mingw32/include/crtdefs.h:138:20: note: previous declaration of 'time_t' was here
 typedef __time64_t time_t;
                    ^
Makefile:383: recipe for target 'lib_a-sysclose.o' failed
make[5]: *** [lib_a-sysclose.o] Error 1
```

**Comment by dscho** [comment]
@mingwandroid thanks!

For the record, these tests are failing in Git's regression test suite when building without any of these patches:

```
t0003-attributes
t0006-date
t0060-path-utils
t1100-commit-tree-options
t1400-update-ref
t1402-check-ref-format
t1501-worktree
t1504-ceiling-dirs
t1506-rev-parse-diagnosis
t1508-at-combinations
t1513-rev-parse-prefix
t2027-checkout-long-paths
t3001-ls-files-others-exclude
t3070-wildmatch
t3200-branch
t3301-notes
t3400-rebase
t3404-rebase-interactive
t3703-add-magic-pathspec
t4150-am
t4208-log-magic-pathspec
t4211-line-log
t5000-tar-tree
t5500-fetch-pack
t5510-fetch
t5601-clone
t5602-clone-remote-exec
t5603-clone-dirname
t5801-remote-helpers
t6006-rev-list-format
t6018-rev-list-glob
t6130-pathspec-noglob
t6132-pathspec-exclude
t6300-for-each-ref
t7201-co
t7400-submodule-basic
t7501-commit
t7502-commit
t8002-blame
t8006-blame-textconv
t9001-send-email
t9350-fast-export
t9700-perl-git
t9903-bash-prompt
```

**Comment by mingwandroid** [comment]
@dscho I've been looking at this again a bit today .. so .. what are your thoughts on the discussion at `http://thread.gmane.org/gmane.comp.version-control.git/278971` in particular:
`http://article.gmane.org/gmane.comp.version-control.git/279004`:

```
Making sure that all the callers expect the same kind of normalization would be a lot of work but I do think that is the best approach in the long run.
```

.. do you know if any work was done for this since or if anyone is planning to work on it? Is it something you would have time to do? I'm not in a position to tackle it at present.

Otherwise, I've got some patches nearly ready for initial comments and testing that makes the t0060-path-utils.sh testsuite results the same on MSYS2 as on Git-for-Windows SDK; I'm well aware that making testsuite results the same is not the correct development strategy unless the changes behind it can be backed up with solid reasoning and that's where I need your critical eye (and have my doubts). I think the correct approach is to fix the normalization code and all the call-sites instead.

Cheers!

**Comment by mingwandroid** [comment]
@dscho Here you go, a gist with a preview of my patch (still contains fprintfs for debugging purposes), the commit msg detailing my logic  https://gist.github.com/mingwandroid/4de33791da01278b5196

.. it also contains the output from t0060, one from Git for Windows SDK and one from MSYS2. Interestingly the error count is the same (11 on each side) but with slightly different failure patterns.

Anyway, let me know what you think. I'll run full testsuites next.

**Comment by dscho** [comment]
> do you know if any work was done for this since or if anyone is planning to work on it?

There has been no work on this that I know of, nor anybody stepping up in the meantime.

As to your strategy, I would like to go a different route: instead of aligning the test results, I would like to question the tests themselves. Why are they (i.e. the failing ones) there, what regressions do they prevent? If there are any, we need to fix MSys2. If they do not prevent any regressions, we need to relax the tests and fix Git for Windows' forked MSys2 runtime.

**Comment by dscho** [comment]
> a gist with a preview of my patch (still contains fprintfs for debugging purposes), the commit msg detailing my logic https://gist.github.com/mingwandroid/4de33791da01278b5196

I miss the part where `GIT_RUNNING_ANCESTOR_TEST` is actually used ;-)

Besides, my thinking goes more into the direction of

```diff
diff --git a/t/t0060-path-utils.sh b/t/t0060-path-utils.sh
index f0152a7..807f5fb 100755
--- a/t/t0060-path-utils.sh
+++ b/t/t0060-path-utils.sh
@@ -7,6 +7,11 @@ test_description='Test various path utilities'
 
 . ./test-lib.sh
 
+# TODO: explain why
+root_ancestor_length=0
+! test_have_prereq MINGW ||
+case "$(test-path-utils print_path /)" in ?*/) root_ancestor_length=-1;; esac
+
 norm_path() {
 	expected=$(test-path-utils print_path "$2")
 	test_expect_success $3 "normalize path: $1 => $2" \
@@ -112,30 +117,30 @@ norm_path /d1/.../d2 /d1/.../d2
 norm_path /d1/..././../d2 /d1/d2
 
 ancestor / / -1
-ancestor /foo / 0
+ancestor /foo / $root_ancestor_length
 ancestor /foo /fo -1
 ancestor /foo /foo -1
 ancestor /foo /bar -1
 ancestor /foo /foo/bar -1
 ancestor /foo /foo:/bar -1
-ancestor /foo /:/foo:/bar 0
-ancestor /foo /foo:/:/bar 0
-ancestor /foo /:/bar:/foo 0
-ancestor /foo/bar / 0
+ancestor /foo /:/foo:/bar $root_ancestor_length
+ancestor /foo /foo:/:/bar $root_ancestor_length
+ancestor /foo /:/bar:/foo $root_ancestor_length
+ancestor /foo/bar / $root_ancestor_length
 ancestor /foo/bar /fo -1
-ancestor /foo/bar /foo 4
+ancestor /foo/bar /foo $((4+$root_ancestor_length))
 ancestor /foo/bar /foo/ba -1
-ancestor /foo/bar /:/fo 0
-ancestor /foo/bar /foo:/foo/ba 4
+ancestor /foo/bar /:/fo $root_ancestor_length
+ancestor /foo/bar /foo:/foo/ba $((4+$root_ancestor_length))
 ancestor /foo/bar /bar -1
 ancestor /foo/bar /fo -1
-ancestor /foo/bar /foo:/bar 4
-ancestor /foo/bar /:/foo:/bar 4
-ancestor /foo/bar /foo:/:/bar 4
-ancestor /foo/bar /:/bar:/fo 0
-ancestor /foo/bar /:/bar 0
-ancestor /foo/bar /foo 4
-ancestor /foo/bar /foo:/bar 4
+ancestor /foo/bar /foo:/bar $((4+$root_ancestor_length))
+ancestor /foo/bar /:/foo:/bar $((4+$root_ancestor_length))
+ancestor /foo/bar /foo:/:/bar $((4+$root_ancestor_length))
+ancestor /foo/bar /:/bar:/fo $root_ancestor_length
+ancestor /foo/bar /:/bar $root_ancestor_length
+ancestor /foo/bar /foo $((4+$root_ancestor_length))
+ancestor /foo/bar /foo:/bar $((4+$root_ancestor_length))
 ancestor /foo/bar /bar -1
 
 test_expect_success 'strip_path_suffix' '
```

But this needs a careful analysis what regressions could be introduced by the change of behavior relative to Git for Windows' current one.

**Comment by mingwandroid** [comment]
Regarding using `GIT_RUNNING_ANCESTOR_TEST `, you are right! I did have it in there but I must have forgot to `git add -p` that bit ;-). Anyway, it was just an extra `&&` condition to the `if` inside the #ifdef _WIN32 block to limit those two blocks from being run to specifically during this ancestor test. I hand-edited my gist.

But I agree that your approach is cleaner as it doesn't leak (more?) MSYS2 shell oddities into the C code, note I didn't test it as the test-suite is still running!

**Comment by mingwandroid** [comment]
.. I should add that while your approach is cleaner, I'm not sure it's coping explicitly with the reality of the C code ending up with paths ending in '/' that are not the root folder whereas mine is, iff `GIT_RUNNING_ANCESTOR_TEST`.

As I said, I didn't test it yet and if we get the same results then great, I definitely prefer your way.

**Comment by mingwandroid** [comment]
`# passed all 150 test(s)`

**Comment by dscho** [comment]
> # passed all 150 test(s)

Yeah!

**Comment by dscho** [comment]
Okay, here they are, the updated patches. With these changed (and the patch to t0025 I proposed), the entire Git test suite passes.

Please note: I am eager to hear how to improve these patches. I am really more interested in getting to a point where Git's test suite passes with MSys2 rather than Git for Windows' fork thereof. So whatever I can do to get there, I am glad to do.

**Comment by mingwandroid** [comment]
Great. For the path conversion changes, can we have a pull request to add each to https://github.com/Alexpux/path_convert ?

**Comment by dscho** [comment]
> can we have a pull request to add each to https://github.com/Alexpux/path_convert ?

@mingwandroid  good point. I pushed [a branch](https://github.com/Alexpux/path_convert/compare/master...dscho:git-for-windows) to my fork. Unfortunately, 50 tests fail, and some of them point out problems with the patches (e.g. the `scp`-style patch catches also some legitimate path list arguments).

I will work on those and keep you posted.

**Comment by mingwandroid** [comment]
@dscho we'll add you as a developer to MSYS2 on Sourceforge if that's OK by you? What's you SF user name?

**Comment by dscho** [comment]
> What's your SF user name?

dscho

:smile:

**Comment by mingwandroid** [comment]
Ok, done, can I please ask that you refer to `MSYS2` as `MSYS2` on our official sites instead of `MSys2` though ;-)

**Comment by dscho** [comment]
> refer to MSYS2 as MSYS2 on our official sites instead of MSys2

Will do.

However, "Create Page" gives me this:

```
Error 403
Admin access required
```

**Comment by mingwandroid** [comment]
That's strange, I added you as a developer and developers have create privileges.

I created a page: https://sourceforge.net/p/msys2/wiki/Path%20Conversion%20Rules/

Can you try to edit it?

**Comment by dscho** [comment]
Okay, I can edit it.
